### DISCUSSION TITLE
Competitive Analysis of IPC: Scaling solutions

### DISCUSSION THREAD

**Discussion by adlrocha** [consensus-shipyard/consensuslab#167] [General]
The goal of this discussions is to source other projects that are focused on building scaling solutions in order to understand how their approaches differ or are similar to IPC. This can really help us evaluate what others are doing, the scale they are targeting (and reaching), and get inspiration of ways to improve our solution or collaborate. 

## Edit (2022-12-19)
In order to have an up-to-date list of projects and their new developments, we are collecting them [in this sheet](https://docs.google.com/spreadsheets/d/1u9e7gXN6pOKuXLHlqzIPO9wWATbShPeU127b-1k0TRg/edit?usp=sharing). Feel free to engage in the discussion and contribute to the analysis dropping a message here, or directly commenting/giving your input in the linked sheet. :pray: 

### Contributions welcome :raised_hands: 
If you come across any other project focused on blockchain scalability, feel free to add a brief overview in this discussion, or just let us know so we can have a look at it and share a brief overview here :pray: 

-------------------
Original description of the issue moved here: https://github.com/protocol/ConsensusLab/discussions/167#discussioncomment-4449399

**Comment by dnkolegov** [comment]
# Avalanche Subnets

> A Subnet is a sovereign network which defines its own rules regarding its membership and token economics. It is composed of a dynamic subset of Avalanche validators working together to achieve consensus on the state of one or more blockchains. Each blockchain is validated by exactly one Subnet, and a Subnet can have many blockchains. A validator may be a member of many Subnets.

At present, subnets are permissioned.

References:

- [Avalanche Subnets Overview](https://docs.avax.network/subnets)

**Reply by adlrocha** [reply]
I guess another big difference with our approach is that they don't support different consensus algorithms in subnet. Avalanche subnets run the same consensus in every network. Also, while our subnets are "strictly" permissioned (you need to stake some collateral to participate from the subnet), they don't need any kind of KYC (which seems to be the case for for Avalanche subnets). 

What I couldn' find are performance numbers for Avalanche to have a sense of their order of magnitude. Let's see if we can find something.

**Comment by adlrocha** [comment]
@aakoshh brought up another project worth considering for this analysis, [Aptos](https://aptoslabs.com/). Like Mysten Labs, the team is founded by former members of Diem/Libra.

> Aptos is a proposed Layer 1 blockchain that uses the [Move programming language](https://101blockchains.com/move-programming-language-tutorial/). The cryptocurrency promises users increased scalability, reliability, security and usability.

Regarding performance:
> Early reports indicate that the Aptos blockchain outperforms [Avalanche](https://learn.bybit.com/altcoins/what-is-avalanche-avax/) under optimal testing conditions. Avalanche can finalize a transaction about 0.7 seconds faster than Aptos. However, in real-world conditions, Avalanche's time to finality is usually about 0.3 seconds slower than Aptos. When it comes to TPS, Aptos blows Avalanche out of the water. Avalanche manages around 4,500 per subnet while the Aptos blockchain achieves [between 130,000–160,000](https://pontem.network/posts/which-blockchain-is-the-fastest).

[Good reference](https://learn.bybit.com/altcoins/what-is-aptos/)

**Comment by adlrocha** [comment]
A new solution worth keeping on our radar: https://www.espressosys.com/
This is their Github: https://github.com/EspressoSystems/espresso
And they seem to be using HotStuff under the hood: https://www.espressosys.com/blog/espresso-hotshot-consensus-designed-for-rollups

**Comment by adlrocha** [comment]
After a quick search, the two immediate projects I could find focusing on similar scaling proposals are.

### Dfinity ICP Subnets
The best description of what Dfinity subnets are is [this reddit thread](https://www.reddit.com/r/dfinity/comments/pjvsp4/why_are_subnets_used_in_dfinity_are_they_similar/).

> "A subnet is (as stupid simple as it sounds) a smaller part of the network, literally a subnetwork as you would expect from traditional computer networks. The name matches pretty clearly to what it is.
>
> The surprising and more helpful part about a subnet that is different from other blockchain projects is that each subnet is a collection of nodes who host smart contracts and perform consensus among each other. You could think of this as each subnet has its own blockchain. These "islands of consensus" communicate via each other and can confirm that messages were validated via Chain Key cryptography. In short, no node or set of nodes could maliciously pretend to be the "subnet with the smart contract canister X". This means each subnet is part of the Internet Computer. No roll-up. No off-chain computing, since it's all "on-chain."
>
> The design intent is to create a computing environment that can scale infinitely. This means the IC should be able to add more subnets without loss of performance and in a trivial manner. Each subnet brings more compute and storage capability and because of Chain Key cryptography, it can be trivial to validate the computation from subnets.

Dinifity's concept of subnets is really close to ours. [This dashboard](https://dashboard.internetcomputer.org/subnets) shows the current subnets deployed in the network and the number of node machines. Their cross-net communication approach is based in a public key based generated through threshold cryptography for the whole subnet. Messages are exchanged individually between subnets (there doesn't seem to be any batching). 

Looking at their performance numbers [here](https://forum.dfinity.org/t/internet-computer-performance-dec-1-2021-load-testing/9240) they claim to be able to achieve _"11,000 updates/second and 230,000 queries per second"_. The results look good but is quite far to the target we are trying to reach. 

### Sui
Sui is another project implementing really interesting scalability ideas. In this case, the key innovation comes from their consensus algorithm and the way they handle state in the blockchain. Sui scales by:
- Using casual order instead total order to execute transactions in parallel when they don't touch shared state.
- This is enabled by using the Move programming language. Move treats objects in the blockchain as first-class, and its semantics are built to ease the detection of shared state and the use of casual ordering. _"Each object value is mutable and owned by an account address at the time of its creation, but can subsequently be frozen and become permanently immutable, or be shared and thus become accessible by other addresses."_

More info about their design can be found in the [docs](https://docs.sui.io). Worth reading the [Sui compared section](https://docs.sui.io/devnet/learn/sui-compared).

I couldn't find any throughput numbers, except to a brief mention to the _130.000 transactions per second in a WAN_ from the Narwhal and Tusk paper. Their [explorer shows](https://explorer.devnet.sui.io/) ~16,5k nodes at the time or writing (I don't know to what extent all of them are validators and participate in the consensus)
